### ISSUE TITLE
[RFC/Codegen] Missing in-place typed writers on slice<u8> (set_i64_le) and restriction on owned field replacement in structs

### ISSUE THREAD

**Issue by sanohiro** [sanohiro/align#1048]
# [RFC/Bug] Missing in-place typed writers on `slice<u8>` (`set_i64_le`) and restriction on owned field replacement (`array<T>`) in structs

## Summary

During optimization work on `align-llm`'s KV cache prefix matching and session reuse (`Session.prefix`), two significant language/compiler limitations in Align were identified that degrade runtime performance by forcing developers to write byte-by-byte manual decomposition loops:

1. **Missing typed in-place writer APIs on `slice<u8>`**: While `slice<u8>` provides binary decoding (`slice.u32_le`, `slice.i64_le`, etc.) and `buffer` provides append-only encoding (`buffer.put_i64_le`), there is no in-place random-access typed writer (e.g. `slice.set_i64_le(offset, value)`).
2. **LLVM Store Merging blocked by per-byte bounds check guards**: When decomposing a 64-bit integer into 8 bytes via `dest[base + i] = ((val >> shift) & 255) as u8`, LLVM emits 60+ instructions including 8 separate bounds check branches (`align_rt_bounds_fail`) and cannot merge the 8 byte stores into a single 64-bit store (`str xN`).
3. **Severe limitation on owned field replacement in structs**: Attempting to store `Session.prefix_ids: array<i64>` and update it via `session.prefix_ids = builder.build()` is rejected by the compiler (`field replacement of array<i64> is not supported yet (owned field replacement currently supports only string and Option<string> leaves; replace the whole struct)`). Attempting whole-struct replacement on a `borrow mut` parameter fails due to borrow movement restrictions.

---

## 1. Disassembly Evidence: 60+ Instructions for a Single 64-Bit Store

In `align-llm/src/runtime_generation.align`, `publish_prefix` updates the cached token prefix in `Session.prefix: buffer`:

```align
fn write_i64_le(borrow mut dest: slice<u8>, base: i64, value: i64) {
  dest[base] = (value & 255) as u8
  dest[base + 1] = ((value >> 8) & 255) as u8
  dest[base + 2] = ((value >> 16) & 255) as u8
  dest[base + 3] = ((value >> 24) & 255) as u8
  dest[base + 4] = ((value >> 32) & 255) as u8
  dest[base + 5] = ((value >> 40) & 255) as u8
  dest[base + 6] = ((value >> 48) & 255) as u8
  dest[base + 7] = ((value >> 56) & 255) as u8
}
```

Disassembly of the compiled binary (`otool -tvV bin/main` on Apple Silicon ARM64):

```assembly
00000001000d68f8  ldr   x9, [x15]
00000001000d68fc  add   x8, x0, #0x1
00000001000d6900  cmp   x8, x1
00000001000d6904  ldr   x9, [x9, x14, lsl #3]
00000001000d6908  strb  w9, [x12, x0]
00000001000d690c  b.ge  0x1000d69d0        ; Bounds check #1 -> abort
00000001000d6910  lsr   x15, x9, #8
00000001000d6914  add   x8, x0, #0x2
00000001000d6918  add   x14, x12, x0
00000001000d691c  cmp   x8, x1
00000001000d6920  strb  w15, [x14, #0x1]
00000001000d6924  b.hs  0x1000d69d0        ; Bounds check #2 -> abort
00000001000d6928  lsr   x15, x9, #16
00000001000d692c  add   x8, x0, #0x3
00000001000d6930  cmp   x8, x1
00000001000d6934  strb  w15, [x14, #0x2]
00000001000d6938  b.hs  0x1000d69d0        ; Bounds check #3 -> abort
... (repeated for all 8 bytes)
```

### Why LLVM cannot optimize this:
Because each `dest[base + i] = ...` performs an individual bounds check that branches to an abort/panic handler, intermediate stores are observable side-effects before subsequent checks. LLVM cannot reorder or coalesce the 8 `strb` instructions into a single `str x9, [x12, x0]`.

For a 2,048-token context, publishing the prefix requires **over 120,000 instructions** instead of ~2,000 instructions.

---

## 2. Gap A: Missing In-Place Typed Writers on `slice<u8>`

Currently, Align provides:
- Binary Reads on `slice<u8>`:
  - `slice.u8(offset)`, `slice.i16_le(offset)`, `slice.u32_le(offset)`, `slice.i64_le(offset)`, `slice.f32_le(offset)`, `slice.f64_le(offset)`
- Binary Writes on `buffer`:
  - `buffer.put_u8(v)`, `buffer.put_u32_le(v)`, `buffer.put_i64_le(v)`, etc. (append-only)

However, there are **no in-place mutating methods** on `slice<u8>` or `buffer`. If a buffer has already been allocated or populated, updating an entry requires manual byte slicing.

### Proposed Solution:
Add mutating methods on `slice<u8>`:
- `slice.set_u16_le(offset: i64, value: u16)`
- `slice.set_u32_le(offset: i64, value: u32)`
- `slice.set_i64_le(offset: i64, value: i64)`
- `slice.set_f32_le(offset: i64, value: f32)`
- `slice.set_f64_le(offset: i64, value: f64)`

Implementation in lowering:
1. Verify `0 <= offset && offset + width <= slice.len` (single bounds check).
2. Emit an unaligned store instruction:
   - On ARM64: `str wN, [ptr, offset]` (for 32-bit) or `str xN, [ptr, offset]` (for 64-bit).
   - On x86-64: `mov dword ptr [ptr + offset], reg` / `mov qword ptr [ptr + offset], reg`.

---

## 3. Gap B: Owned Field Replacement Restriction (`array<T>`)

In `crates/align_sema/src/lib.rs:47170`:
```rust
// Direct field replacement has typed drop-old lowering only for the two
// scalar owned types: `string` and `Option<string>`.
"field replacement of {} is not supported yet (owned field replacement currently supports only `string` and `Option<string>` leaves; replace the whole struct)"
```

When designing stateful objects (like `Session`), developers naturally want:
```align
pub Session {
  ...
  prefix_ids: array<i64>,
}

fn publish_prefix(borrow mut session: Session, ...) {
  session.prefix_ids = new_ids  // ERROR!
}
```

Because `array<i64>` field replacement is rejected:
1. The compiler suggests "replace the whole struct".
2. But in `borrow mut session: Session`, `session` cannot move out its other owned fields (e.g. `prefill_slots: buffer`):
   ```
   error: cannot move a field out of borrowed parameter 'session'
   error: moving a nested struct field out of a struct is not supported yet — clone it, or move the whole struct
   ```
3. Therefore, developers are forced to use `Session.prefix: buffer` and modify it via `bytes()`, falling into the 60-instruction trap described in Gap A.

### Proposed Solution:
Support typed drop-old lowering for `array<T>` and other owned types in struct field assignments (`align_sema/src/lib.rs:47170`).

---

## 4. Gap C: Auto-Vectorization (NEON/AVX) for `greedy` / Argmax Loops

In LLM token generation, `greedy` selection executes over 151,936 logits (Qwen2 vocabulary):
```align
fn greedy(borrow logits: slice<u8>) -> Result<i64, Error> {
  ...
  loop {
    if token >= n_vocab { break }
    bits := logits.u32_le(offset)
    if (bits & 2139095040) == 2139095040 { return Err(Error.Invalid) }
    value := logits.f32_le(offset)
    if value > best_value {
      best = token
      best_value = value
    }
    ...
  }
}
```
Currently, this emits a pure scalar loop (approx. 1.5 million scalar instructions per decode step).
The early error check (`(bits & 2139095040) == 2139095040`) prevents LLVM from auto-vectorizing with ARM64 NEON (`fmaxnm` / `fmaxnmv`).

Providing standard library SIMD primitives or an intrinsic `slice<u8>.argmax_f32_le() -> Result<i64, Error>` would yield a 4x–8x speedup on CPU argmax during decode.

**Comment by sanohiro** [comment]
The audit confirms the missing typed in-place writers and owning-array field replacement. Plan 65 accepts the complete endian scalar-store family and replacement of already-admitted owning fields with complete Drop, together with plan 61 access/provenance enforcement. These surfaces are still in progress.

Each new typed store will validate its full width before writing; this is not permission to merge existing byte stores across later failing checks. Ordered argmax SIMD is tracked separately because nonfinite rejection, tie order and IEEE behavior must remain exact. #1056 already supplies float inspection and stronger byte-loop proofs, but does not close these remaining writer/replacement/SIMD findings.

**Comment by sanohiro** [comment]
Typed slice writers and owned field replacement implemented in branch `feat/array-truncate` (commit `eca0cf62`).

- `slice<u8>.set_S_E(offset: i64, value: S) -> ()` for all scalars (`u8`, `i8`, `u16`, `i16`, `u32`, `i32`, `u64`, `i64`, `f32`, `f64`) and endian suffixes (`le`, `be`), validating atomic bounds and enforcing mutable receiver places.
- General owned struct field replacement (`owner.field_path = value`) for admitted owning types (`array<T>`, `buffer`, `string`, `Option<T>`), evaluating RHS first, dropping old live destination once, and updating in-place without whole-struct replacement.

SIMD argmax remains tracked separately. Verified across focused owner tests (`tests/bytes_ops.rs`, `tests/owned_field_replacement.rs`), bounded gate, and workspace Clippy. Ready for PR.
